internal/manifest/manifest.go

```diff
@@ -0,0 +1,245 @@
+	Mode      string   `json:"mode"`
+	Slices    []string `json:"slices"`
+	Hash      string   `json:"sha256,omitempty"`
+	FinalHash string   `json:"final_sha256,omitempty"`
+	Size      uint64   `json:"size,omitempty"`
+	Link      string   `json:"link,omitempty"`
+}
+
+type Content struct {
+	Kind  string `json:"kind"`
+	Slice string `json:"slice"`
+	Path  string `json:"path"`
+}
+
+type Manifest struct {
+	db *jsonwall.DB
+}
+
+func Read(absPath string) (manifest *Manifest, err error) {
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("cannot read manifest: %s", err)
+		}
+	}()
+
+	file, err := os.OpenFile(absPath, os.O_RDONLY, 0644)
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+	r, err := zstd.NewReader(file)
+	if err != nil {
+		return nil, err
+	}
+	defer r.Close()
+	jsonwallDB, err := jsonwall.ReadDB(r)
+	if err != nil {
+		return nil, err
+	}
+
+	manifest = &Manifest{db: jsonwallDB}
+	err = validate(manifest)
+	if err != nil {
+		return nil, err
+	}
+	return manifest, nil
+}
+
+func (manifest *Manifest) IteratePath(prefix string, f func(Path) error) (err error) {
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("cannot read manifest: %s", err)
+		}
+	}()
+
+	iter, err := manifest.db.IteratePrefix(map[string]string{"kind": "path", "path": prefix})
+	if err != nil {
+		return err
+	}
+	for iter.Next() {
+		var path Path
+		err := iter.Get(&path)
+		if err != nil {
+			return err
+		}
+		err = f(path)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (manifest *Manifest) IteratePkgs(f func(Package) error) (err error) {
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("cannot read manifest: %s", err)
+		}
+	}()
+
+	iter, err := manifest.db.Iterate(map[string]string{"kind": "package"})
+	if err != nil {
+		return err
+	}
+	for iter.Next() {
+		var pkg Package
+		err := iter.Get(&pkg)
+		if err != nil {
+			return err
+		}
+		err = f(pkg)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (manifest *Manifest) IterateSlices(pkgName string, f func(Slice) error) (err error) {
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("cannot read manifest: %s", err)
+		}
+	}()
+
+	iter, err := manifest.db.IteratePrefix(map[string]string{"kind": "slice", "name": pkgName})
+	if err != nil {
+		return err
+	}
+	for iter.Next() {
+		var slice Slice
+		err := iter.Get(&slice)
+		if err != nil {
+			return err
+		}
+		err = f(slice)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func validate(manifest *Manifest) (err error) {
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf(`invalid manifest: %s`, err)
+		}
+	}()
+
+	pkgExist := map[string]bool{}
+	iter, err := manifest.db.Iterate(map[string]string{"kind": "package"})
+	if err != nil {
+		return err
+	}
+	for iter.Next() {
+		var pkg Package
+		err := iter.Get(&pkg)
+		if err != nil {
+			return err
+		}
+		if pkg.Kind != "package" {
+			return fmt.Errorf(`in packages expected kind "package", got %q`, pkg.Kind)
```

Comment:
That one seems a bit suspect. This is testing the jsonwall package instead of the file being validated. It would make more sense to check if the name exists, or is duplicated, for example, since it's about to put it into a map a lose that information.

Reply:
I agree and I have removed the checks in https://github.com/canonical/chisel/pull/144/commits/e2c2645e38e0c0d31f7c435970f4376cc55d0bc7.

About the duplicated part, I cannot find a place in the spec or in the jsonwall format documentation where it states that duplicates are not valid, hence why I was not checking for them. Do you think we should throw an error or just ignore it?

Reply:
EDIT: Discussed offline, it will be "undefined behavior" for now. We consider the jsonwall to be malformed when it contains duplicated entries and we make no promise to make sense of it.